### PR TITLE
Initial documentation.

### PR DESCRIPTION
**Pull request by mblondel**
To do in separate pull requests:
- example gallery (@fabianp)
- doctests
- consistency improvements
- github actions for deploying the doc?

### CODE REVIEW

File: docs/basics.rst [hunk 1]
```diff
@@ -0,0 +1,21 @@
+Basics
+======
+
+This section describes useful concepts across all JAXopt.
+
+Pytrees
+-------
+
+Pytrees are an essential concept in JAX and JAXopt. They can be thought as a
+generalization of vector. They are a way to structure parameters or weights
```
Comment (fabianp) [inline docs/basics.rst:10]: vector-> vectors, or of -> of a

File: docs/basics.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+using tuples and dictionaries. Many solvers in JAXopt have native support for
+pytrees.
+
+Double precision
+----------------
+
+JAX uses 32-bit floats by default. Howver, for some algorithms, this is not
```
Comment (fabianp) [inline docs/basics.rst:17]: floats -> floating point precision, Howver -> However

File: docs/basics.rst [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+enough. Double precision can be enabled by adding the following at the
```
Comment (fabianp) [inline docs/basics.rst:18]: I would add (64 bits) after precision since before precision was started in bits

File: docs/constrained.rst [hunk 4]
```diff
@@ -0,0 +1,126 @@
+Constrained optimization
+========================
+
+This section is concerned with problems of the form
+
+.. math::
+
+    \min_{x} f(x, \theta) \textrm{ subject to } x \in \mathcal{C}(\upsilon),
+
+where :math:`f(x, \theta)` is a smooth function, :math:`x` are the parameters
```
Comment (fabianp) [inline docs/constrained.rst:10]: I hope people use it on things like relu that are not technically smooth. I would replace smooth with "functions for which we can compute its gradient"

File: docs/constrained.rst [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+with respect to which the function is minimized, :math:`\theta` are optional
+extra arguments, :math:`\mathcal{C}(\upsilon)` is a convex set
```
Comment (fabianp) [inline docs/constrained.rst:12]: optional extra seems redundant, I would use either optional or extra

File: docs/constrained.rst [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    jaxopt.projection.projection_l2_sphere
+    jaxopt.projection.projection_l2_ball
+    jaxopt.projection.projection_linf_ball
+    jaxopt.projection.projection_hyperplane
+    jaxopt.projection.projection_halfspace
+    jaxopt.projection.projection_affine_set
+    jaxopt.projection.projection_polyhedron
+    jaxopt.projection.projection_box_section
+
+All projections always have two arguments: the input to be projected and the
```
Comment (fabianp) [inline docs/constrained.rst:50]: All is unnecessary

File: docs/non_smooth.rst [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+Proximal gradient
+-----------------
+
+.. autosummary::
+  :toctree: _autosummary
+
+    jaxopt.ProximalGradient
+
+
+Block Coordinate Descent
```
Comment (fabianp) [inline docs/non_smooth.rst:26]: Most sections don't  capitalize each word, make consistent

File: docs/non_smooth.rst [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+.. autosummary::
+  :toctree: _autosummary
+
+    jaxopt.BlockCoordinateDescent
+
+Proximal operators
+------------------
+
+Proximal gradient and block coordinate descent do not access :math:`g(x, \upsilon)`
+directly but instead requires its associated proximal operator
```
Comment (fabianp) [inline docs/non_smooth.rst:38]: requires-> require

File: docs/stochastic.rst [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+where :math:`f(x, \theta, D)` is a smooth function, :math:`x` are the
+parameters with respect to which the function is minimized, :math:`\theta` are
+optional fixed extra arguments and :math:`D` is a random variable
+(typically a mini-batch).
+
+Defining an objective function
+------------------------------
+
+Objective functions must contain a compulsory ``data`` argument corresponding to :math:`D` above.
```
Comment (fabianp) [inline docs/stochastic.rst:18]: must and compulsory is redundant

File: docs/constrained.rst (hunk 4 above)
Comment (fabianp) [reply docs/constrained.rst:10]: Same for other files

**Review by fabianp — APPROVED** [review]
Great job. Just cosmetic comments

**Comment by mblondel** [comment]
Thanks a lot @fabianp! I addressed your comments and force-pushed.

**Pull request merged**